File: movtools-server/src/Movtools.Server.Domain/Entities/Lens.cs
namespace Movtools.Server.Domain.Entities;

/// <summary>
/// 镜头实体
/// </summary>
public sealed class Lens : EntityBase
{
    /// <summary>
    /// 镜头代码（唯一标识）
    /// </summary>
    public string Code { get; set; } = string.Empty;

    /// <summary>
    /// 镜头名称
    /// </summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>
    /// 所属剧集ID
    /// </summary>
    public Guid EpisodeId { get; set; }

    /// <summary>
    /// 所属剧集实体
    /// </summary>
    public Episode Episode { get; set; } = null!;

    /// <summary>
    /// 当前状态
    /// </summary>
    public string Status { get; set; } = LensStatuses.Wip;

    /// <summary>
    /// 序号（用于排序）
    /// </summary>
    public int Sequence { get; set; }

    /// <summary>
    /// 单镜头帧数
    /// </summary>
    public int SingleFrame { get; set; }

    /// <summary>
    /// 镜头描述
    /// </summary>
    public string? Description { get; set; }

    /// <summary>
    /// 制作人员
    /// </summary>
    public string? Maker { get; set; }

    /// <summary>
    /// 正式负责人账号ID
    /// </summary>
    public Guid? MakerUserId { get; set; }

    /// <summary>
    /// 制作人员原始姓名
    /// </summary>
    public string? MakerNameRaw { get; set; }

    /// <summary>
    /// 制作人员匹配状态
    /// </summary>
    public string? MakerMatchStatus { get; set; }

    /// <summary>
    /// 根镜头代码
    /// </summary>
    public string? RootCode { get; set; }

    /// <summary>
    /// 逻辑路径
    /// </summary>
    public string? LogicalPath { get; set; }

    /// <summary>
    /// 版本标签
    /// </summary>
    public string? VersionTag { get; set; }

    /// <summary>
    /// 当前版本号
    /// </summary>
    public string? VersionNum { get; set; } = "V01";

    /// <summary>
    /// 内部审片状态码
    /// </summary>
    public string InternalReviewStatusCode { get; set; } = LensInternalReviewStatuses.NotInReview;

    /// <summary>
    /// 内部审片状态更新时间
    /// </summary>
    public DateTimeOffset? InternalReviewUpdatedAtUtc { get; set; }

    /// <summary>
    /// 最近审片任务ID
    /// </summary>
    public Guid? LatestReviewTaskId { get; set; }

    /// <summary>
    /// 最近导演反馈时间
    /// </summary>
    public DateTimeOffset? LatestDirectorFeedbackAtUtc { get; set; }

    /// <summary>
    /// 待处理导演反馈数量
    /// </summary>
    public int PendingDirectorFeedbackCount { get; set; }

    /// <summary>
    /// 布局标签
    /// </summary>
    public string? LayoutTag { get; set; }

    /// <summary>
    /// 备注
    /// </summary>
    public string? Comment { get; set; }

    /// <summary>
    /// 是否已归档
    /// </summary>
    public bool IsArchived { get; set; }

    /// <summary>
    /// 行版本（用于并发控制）
    /// </summary>
    public long RowVersion { get; set; }

    /// <summary>
    /// 状态历史记录集合
    /// </summary>
    public ICollection<LensStatusHistory> StatusHistories { get; set; } = [];

    /// <summary>
    /// 文件绑定集合
    /// </summary>
    public ICollection<LensFileBinding> FileBindings { get; set; } = [];
}

/// <summary>
/// 镜头状态常量
/// </summary>
public static class LensStatuses
{
    /// <summary>工作中</summary>
    public const string Wip = "WIP";
    /// <summary>已提交</summary>
    public const string Submitted = "SUBMITTED";
    /// <summary>审核中</summary>
    public const string InReview = "IN_REVIEW";
    /// <summary>已通过</summary>
    public const string Approved = "APPROVED";
    /// <summary>返修中</summary>
    public const string Rework = "REWORK";
    /// <summary>已拒绝</summary>
    public const string Rejected = "REJECTED";
    /// <summary>已关闭</summary>
    public const string Closed = "CLOSED";

    /// <summary>
    /// 所有状态列表
    /// </summary>
    public static readonly string[] All = [Wip, Submitted, InReview, Approved, Rework, Rejected, Closed];

    /// <summary>
    /// 验证状态是否有效
    /// </summary>
    public static bool IsValid(string status) => All.Contains(status);

    /// <summary>
    /// 允许的状态转换映射
    /// </summary>
    public static readonly Dictionary<string, string[]> AllowedTransitions = new()
    {
        [Wip] = [Submitted, Closed],
        [Submitted] = [Wip, InReview, Rework],
        [InReview] = [Approved, Rejected, Rework],
        [Approved] = [Rework, Rejected, Closed],
        [Rework] = [Wip, Submitted],
        [Rejected] = [Wip, Submitted],
        [Closed] = [Wip]
    };

    /// <summary>
    /// 检查是否可以从一个状态转换到另一个状态
    /// </summary>
    public static bool CanTransition(string from, string to)
    {
        return AllowedTransitions.TryGetValue(from, out var allowed) && allowed.Contains(to);
    }
}

/// <summary>
/// 镜头制作人员匹配状态常量
/// </summary>
public static class LensMakerMatchStatuses
{
    public const string Matched = "matched";
    public const string Unmatched = "unmatched";
    public const string Unassigned = "unassigned";

    public static readonly string[] All = [Matched, Unmatched, Unassigned];

    public static bool IsValid(string? status)
        => !string.IsNullOrWhiteSpace(status) && All.Contains(status.Trim().ToLowerInvariant());

    public static string Normalize(string? status)
        => string.IsNullOrWhiteSpace(status) ? Unassigned : status.Trim().ToLowerInvariant();
}

/// <summary>
/// 镜头内部审片状态常量
/// </summary>
public static class LensInternalReviewStatuses
{
    public const string NotInReview = "NOT_IN_REVIEW";
    public const string ReadyForReview = "READY_FOR_REVIEW";
    public const string InDirectorReview = "IN_DIRECTOR_REVIEW";
    public const string PendingFeedbackFix = "PENDING_FEEDBACK_FIX";
    public const string FixUpdated = "FIX_UPDATED";
    public const string DirectorApproved = "DIRECTOR_APPROVED";

    public static readonly string[] All =
    [NotInReview, ReadyForReview, InDirectorReview, PendingFeedbackFix, FixUpdated, DirectorApproved];

    public static bool IsValid(string? status)
        => !string.IsNullOrWhiteSpace(status) && All.Contains(status.Trim().ToUpperInvariant());

    public static string Normalize(string? status)
        => string.IsNullOrWhiteSpace(status) ? NotInReview : status.Trim().ToUpperInvariant();

        public static bool CanTransition(string from, string to)
        {
            var normalizedFrom = Normalize(from);
            var normalizedTo = Normalize(to);
            return normalizedFrom switch
            {
                NotInReview => normalizedTo is ReadyForReview,
                ReadyForReview => normalizedTo is InDirectorReview,
                InDirectorReview => normalizedTo is PendingFeedbackFix or DirectorApproved,
                PendingFeedbackFix => normalizedTo is FixUpdated,
                FixUpdated => normalizedTo is ReadyForReview,
                DirectorApproved => false,
                _ => false
            };
        }
}

File: movtools-client/src/types/lens.ts
/**
 * 镜头类型定义
 * 
 * 本文件定义镜头相关的所有类型：
 * - 镜头状态、资产类型
 * - 版本绑定、版本问题
 * - 生命周期事件
 * - 列表响应、详情响应
 */
import type { LayoutReferenceCheckRecord, LayoutReferenceCheckStatus, LensLayoutCandidate } from './fileCheck';
import type { InternalReviewStatusCode } from '../lib/internalReview';

/** 镜头制作状态 */
export type LensStatus = '制作' | '提交' | '返修' | '通过' | '关闭';

/** 镜头状态流转动作 */
export type LensStatusAction = 'submit' | 'approve' | 'rework' | 'close';

export type LensRecentStatusAction = LensStatusAction;

/** 镜头资产文件类型 */
export type LensAssetType = 'ma' | 'mov';
export type LensBindingType = 'ma' | 'mov' | 'layout' | 'layoutVideo';
export type MakerMatchStatus = 'matched' | 'unmatched' | 'unassigned';

export interface LensFileBindingSyncRequest {
  bindingType: LensBindingType;
  relativePath: string;
  sourceRoot?: string | null;
  versionNum?: string | null;
  fileName?: string | null;
}

export interface LensFileBindingSyncResponse {
  bindingId: string;
  lensId: string;
  lensCode: string;
  bindingType: LensBindingType;
  relativePath: string;
  sourceRoot?: string | null;
  versionNum?: string | null;
  fileName?: string | null;
  bindTime: string;
}

/** 镜头版本问题描述 */
export interface LensVersionIssue {
  fileType: LensAssetType;
  reason: '未绑定' | '文件缺失' | '多候选待确认' | '帧数不匹配';
  message: string;
  candidatePaths?: string[];
}

/** 镜头版本匹配候选文件 */
export interface LensVersionMatchCandidate {
  fileName: string;
  relativePath: string;
  absolutePath: string;
  sourceRoot: string;
  score: number;
  extractedVersion?: number;
}

/** 镜头版本匹配调试信息 */
export interface LensVersionMatchDebug {
  fileType: LensAssetType;
  versionNum: string;
  scanRoots: string[];
  scannedFileCount: number;
  relatedFileCount: number;
  candidateCount: number;
  note: string;
  relatedFiles: LensVersionMatchCandidate[];
  candidates: LensVersionMatchCandidate[];
}

/** 镜头版本文件绑定记录 */
export interface LensVersionBinding {
  fileId: string;
  bindingId?: string;
  lensId?: string;
  lensCode: string;
  versionNum: string;
  fileType: LensAssetType;
  bindingType?: LensBindingType;
  fileName?: string;
  relativePath: string;
  bindTime: string;
  absolutePath: string;
  exists: boolean;
  sourceRoot?: string;
  mediaPreviewUrl?: string;
  mediaDurationSeconds?: number;
  mediaFrameCount?: number;
  mediaFps?: number;
   mediaWidth?: number;
   mediaHeight?: number;
   mediaCodecName?: string;
  mediaCodecLongName?: string;
  mediaCodecProfile?: string;
  mediaPixelFormat?: string;
  mediaPreviewMode?: 'direct' | 'proxy' | 'pending';
  mediaPreviewNote?: string;
  mediaPreviewProgressPercent?: number;
}

/** 镜头生命周期事件记录 */
export interface LensLifecycleEvent {
  eventId: string;
  lensId: string;
  eventType: '创建' | '基础信息更新' | '状态流转' | '文件绑定';
  title: string;
  detail: string;
  fromStatus?: LensStatus;
  toStatus?: LensStatus;
  versionNum: string;
  fileName: string;
  eventTime: string;
  editable: boolean;
  reworkNote?: string;
  attachments: LensLifecycleAttachment[];
}

export interface LensLifecycleAttachment {
  attachmentId: string;
  eventId: string;
  fileName: string;
  relativePath: string;
  absolutePath: string;
  previewUrl: string;
  createTime: string;
}

export interface ServerRepairAttachmentResponse {
  id: string;
  lensId: string;
  lensStatusHistoryId?: string | null;
  fileName: string;
  originalName: string;
  fileSize: number;
  sortOrder: number;
  previewUrl: string;
  createdAtUtc: string;
}

/** 镜头版本快照 - 包含特定版本的所有信息 */
export interface LensVersionSnapshot {
  versionNum: string;
  fileName: string;
  issues: LensVersionIssue[];
  bindings: LensVersionBinding[];
  matchDebug: Partial<Record<LensAssetType, LensVersionMatchDebug>>;
}

export interface LensRecord {
  lensId: string;
  episodeId: string;
  lensCode: string;       // 来自 Excel "镜头名称"
  sceneNo: number;       // 来自 Excel "场次"
  lensName: string;       // 来自 Excel "镜头名称"
  singleFrame: number;   // 来自 Excel "镜头时长（帧数）"
  maker: string;         // 来自 Excel "负责人"
  makerUserId?: string | null;
  makerNameRaw?: string | null;
  makerDisplayName?: string | null;
  makerMatchStatus?: MakerMatchStatus;
  note?: string;
  lensStatus: LensStatus;
  versionTag: string;
  versionNum: string;
  fileName: string;
  updateTime: string;
  currentVersionIssues: LensVersionIssue[];
  currentVersionReady: boolean;
  currentVersionMatchedFileNames: string[];
  layoutCandidateCount: number;
  selectedLayoutFileName: string;
  selectedLayoutRelativePath: string;
  layoutReady: boolean;
  layoutVideoReady: boolean;
  layoutVideoFileName: string;
  layoutVideoRelativePath: string;
  layoutVideoAbsolutePath: string;
  layoutVideoVersionNum: string;
  layoutVideoPreviewUrl?: string;
  layoutVideoDurationSeconds?: number;
  layoutVideoFrameCount?: number;
  layoutVideoFps?: number;
   layoutVideoWidth?: number;
   layoutVideoHeight?: number;
   layoutVideoCodecName?: string;
  layoutVideoCodecLongName?: string;
  layoutVideoCodecProfile?: string;
  layoutVideoPixelFormat?: string;
  layoutVideoPreviewMode?: 'direct' | 'proxy' | 'pending';
  layoutVideoPreviewNote?: string;
  layoutVideoPreviewProgressPercent?: number;
  recentStatusAction?: LensRecentStatusAction;
  recentStatusActionLabel: string;
  recentStatusActionTime: string;
  layoutReferenceStatus: LayoutReferenceCheckStatus;
  layoutReferenceIssueCount: number;
  layoutReferenceLastCheckTime?: string;
  internalReviewStatusCode?: InternalReviewStatusCode | null;
  internalReviewStatusName?: string | null;
  internalReviewUpdatedAtUtc?: string | null;
  latestReviewTaskId?: string | null;
  latestDirectorFeedbackAtUtc?: string | null;
  pendingDirectorFeedbackCount?: number;
  submissionAllowed?: boolean;
}

export interface LensListResponse {
  success: boolean;
  lenses: LensRecord[];
  autoInitializedLensIds?: string[];
  activeProjectId: string | null;
  activeProjectName?: string;
  activeEpisodeId?: string | null;
  activeEpisodeName?: string;
  activeEpisodeCode?: string;
  episodeVersionTag?: string;
  episodeLayoutTag?: string;
  error?: string;
}

export interface LensDetailPayload {
  lens: LensRecord;
  versions: LensVersionSnapshot[];
  history: LensLifecycleEvent[];
  layoutCandidates: LensLayoutCandidate[];
  serverBindings?: ServerLensFileBindingResponse[];
  layoutReferenceCheck?: LayoutReferenceCheckRecord;
  directorFeedbacks?: LensDirectorFeedback[];
}

export interface ServerRepairAttachmentResponse {
  id: string;
  lensId: string;
  lensStatusHistoryId?: string | null;
  fileName: string;
  originalName: string;
  fileSize: number;
  sortOrder: number;
  previewUrl: string;
  createdAtUtc: string;
}

export interface LensDirectorFeedback {
  feedbackId: string;
  reviewTaskId: string;
  lensId: string;
  lensCode: string;
  versionNum?: string | null;
  frameNumber?: number | null;
  timecode?: string | null;
  commentText?: string | null;
  frameImagePath?: string | null;
  annotatedImagePath?: string | null;
  thumbnailPath?: string | null;
  annotationDataJson?: string | null;
  createdByDisplayName?: string | null;
  createdAtUtc: string;
  updatedAtUtc?: string | null;
  decisionCode?: 'PENDING' | 'CHANGES_REQUIRED' | 'APPROVED' | null;
  decisionName?: string | null;
}

export interface LensDetailResponse {
  success: boolean;
  detail?: LensDetailPayload;
  error?: string;
}

/** 服务端镜头详情响应（直接映射） */
export interface ServerLensDetailResponse {
  lens: ServerLensResponse;
  versions: ServerLensVersionResponse[];
  fileBindings: ServerLensFileBindingResponse[];
  layoutCandidates: ServerLayoutCandidateResponse[];
  currentLayout: ServerLayoutInfoResponse | null;
  layoutReferenceCheck: ServerLayoutReferenceCheckResponse | null;
}

export interface ServerLensResponse {
  id: string;
  code: string;
  name: string;
  episodeId: string;
  status: string;
  sequence: number;
  makerUserId?: string | null;
  makerNameRaw?: string | null;
  makerDisplayName?: string | null;
  makerMatchStatus?: MakerMatchStatus;
  description?: string | null;
  rootCode?: string | null;
  logicalPath?: string | null;
  versionTag?: string | null;
  versionNum?: string;
  layoutTag?: string | null;
  comment?: string | null;
  internalReviewStatusCode?: import('../lib/internalReview').InternalReviewStatusCode | null;
  internalReviewStatusName?: string | null;
  internalReviewUpdatedAtUtc?: string | null;
  latestReviewTaskId?: string | null;
  latestDirectorFeedbackAtUtc?: string | null;
  pendingDirectorFeedbackCount?: number;
  submissionAllowed?: boolean;
  isArchived: boolean;
  rowVersion: number;
  createdAtUtc: string;
  updatedAtUtc: string;
  singleFrame?: number;
}

export interface ServerLensVersionResponse {
  versionNum: string;
  fileName?: string | null;
  logicalPath?: string | null;
  issues: ServerVersionIssueResponse[];
}

export interface ServerVersionIssueResponse {
  issueType: string;
  description: string;
  filePath?: string | null;
}

export interface ServerVersionBindingResponse {
  bindingType: string;
  fileName: string;
  filePath?: string | null;
  isMatched: boolean;
}

export interface ServerLensFileBindingResponse {
  bindingId: string;
  lensId: string;
  lensCode: string;
  bindingType: LensBindingType;
  relativePath: string;
  sourceRoot?: string | null;
  versionNum?: string | null;
  fileName?: string | null;
  bindTime: string;
}

export interface ServerLayoutCandidateResponse {
  fileName: string;
  relativePath: string;
  matchedLensCode?: string | null;
  matchScore: number;
  scannedAt: string;
}

export interface ServerLayoutInfoResponse {
  fileName: string;
  relativePath: string;
  videoFileName?: string | null;
  videoRelativePath?: string | null;
  videoReady: boolean;
  selectedAt: string;
}

export interface ServerLayoutReferenceCheckResponse {
  totalReferences: number;
  validReferences: number;
  missingReferences: number;
  missingReferencePaths: string[];
}
